### PR TITLE
Addition of new resource and data source for L4-L7 devices: aci_l4_l7_devices

### PR DESCRIPTION
**Pull request by shrsr**

### CODE REVIEW

File: aci/data_source_aci_vnsldevvip.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"context_aware": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"devtype": {
```
Comment (lhercot) [inline aci/data_source_aci_vnsldevvip.go:39]: device_type or what is used in UI?

File: aci/data_source_aci_vnsldevvip.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"func_type": {
```
Comment (lhercot) [inline aci/data_source_aci_vnsldevvip.go:44]: function_type or what is used in UI?

File: aci/data_source_aci_vnsldevvip.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"package_model": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"prom_mode": {
```
Comment (lhercot) [inline aci/data_source_aci_vnsldevvip.go:73]: what is it in the UI?

File: aci/data_source_aci_vnsldevvip.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"svc_type": {
```
Comment (lhercot) [inline aci/data_source_aci_vnsldevvip.go:78]: service_type

File: aci/resource_aci_vnsldevvip.go [hunk 5]
```diff
@@ -0,0 +1,587 @@
...
+			"context_aware": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"multi-Context",
+					"single-Context",
+				}, false),
+			},
+			"devtype": {
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:52]: device_type or same as UI

File: aci/resource_aci_vnsldevvip.go [hunk 6]
```diff
@@ -0,0 +1,587 @@
...
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"CLOUD",
+					"PHYSICAL",
+					"VIRTUAL",
+				}, false),
+			},
+			"func_type": {
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:63]: function_type or same as UI

File: aci/resource_aci_vnsldevvip.go [hunk 7]
```diff
@@ -0,0 +1,587 @@
...
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"package_model": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"prom_mode": {
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:111]: What is it in the UI?

File: aci/resource_aci_vnsldevvip.go [hunk 8]
```diff
@@ -0,0 +1,587 @@
...
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"no",
+					"yes",
+				}, false),
+			},
+			"svc_type": {
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:120]: service_type

File: aci/resource_aci_vnsldevvip.go [hunk 9]
```diff
@@ -0,0 +1,587 @@
...
+
+		CustomizeDiff: func(_ context.Context, diff *schema.ResourceDiff, v interface{}) error {
+			// Plan time validation.
+			device_type, _ := diff.GetOk("devtype")
+			_, virtual_ok := diff.GetOk("relation_vns_rs_al_dev_to_dom_p")
+			_, physical_ok := diff.GetOk("relation_vns_rs_al_dev_to_phys_dom_p")
+			if device_type.(string) == "VIRTUAL" && !virtual_ok {
+				return errors.New(`"relation_vns_rs_al_dev_to_dom_p" is required when "devtype" is VIRTUAL`)
+			}
+			if device_type.(string) == "PHYSICAL" && !physical_ok {
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:179]: shouldn't this be an else if?

File: aci/resource_aci_vnsldevvip.go [hunk 10]
```diff
@@ -0,0 +1,587 @@
...
+
+func resourceAciL4ToL7DevicesRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	log.Printf("[DEBUG] %s: Beginning Read", d.Id())
+	aciClient := m.(*client.Client)
+	dn := d.Id()
+
+	vnsLDevVip, err := getRemoteL4ToL7Devices(aciClient, dn)
+	if err != nil {
+		d.SetId("")
+		return diag.FromErr(err)
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:538]: return nil

File: go.mod [hunk 11]
```diff
@@ -3,7 +3,7 @@ module github.com/terraform-providers/terraform-provider-aci
 go 1.12
 
 require (
-	github.com/ciscoecosystem/aci-go-client v1.36.1
+	github.com/ciscoecosystem/aci-go-client v1.37.0
```
Comment (lhercot) [inline go.mod:6]: update to v1.38.0 to fix compile issue.

File: website/docs/d/l4_l7_devices.html.markdown [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+* `tenant_dn` - (Required) Distinguished name of parent Tenant object.
+* `name` - (Required) Name of object L4-L7 Devices.
+
+## Attribute Reference ##
+* `id` - Attribute id set to the Dn of the L4-L7 Devices.
+* `annotation` - (Optional) Annotation of object L4-L7 Devices.
+* `name_alias` - (Optional) Name Alias of object L4-L7 Devices.
+* `active` - (Optional) Active mode. Allowed values are "no", "yes", and default value is "no".
+* `context_aware` - (Optional) Tenancy. A value to determine if the L4-L7 device cluster supports multiple contexts (VRFs). Allowed values are "multi-Context", "single-Context", and default value is "single-Context".
```
Comment (lhercot) [inline website/docs/d/l4_l7_devices.html.markdown:44]: Determine if... (remove Tenancy. A value to)

**Review by lhercot — CHANGES_REQUESTED** [review]

File: aci/provider.go [hunk 13]
```diff
@@ -301,6 +301,7 @@ func Provider() *schema.Provider {
 			"aci_aaa_domain_relationship":                  resourceAciDomainRelationship(),
 			"aci_aaep_to_domain":                           resourceAciInfraRsDomP(),
 			"aci_action_rule_additional_communities":       resourceAciRtctrlSetAddComm(),
+			"aci_l4_l7_devices":                            resourceAciL4ToL7Devices(),
```
Comment (lhercot) [inline aci/provider.go:304]: aci_l4_l7_device

File: aci/provider.go [hunk 14]
```diff
@@ -523,6 +524,7 @@ func Provider() *schema.Provider {
 			"aci_aaa_domain_relationship":                  dataSourceAciDomainRelationship(),
 			"aci_aaep_to_domain":                           dataSourceAciInfraRsDomP(),
 			"aci_action_rule_additional_communities":       dataSourceAciRtctrlSetAddComm(),
+			"aci_l4_l7_devices":                            dataSourceAciL4ToL7Devices(),
```
Comment (lhercot) [inline aci/provider.go:527]: aci_l4_l7_device

File: aci/resource_aci_vnsldevvip.go [hunk 15]
```diff
@@ -0,0 +1,572 @@
...
+			},
+			"relation_vns_rs_al_dev_to_phys_dom_p": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Create relation to phys:DomP",
+			}})),
+
+		CustomizeDiff: func(_ context.Context, diff *schema.ResourceDiff, v interface{}) error {
+			// Plan time validation.
+			device_type, _ := diff.GetOk("device_type")
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:168]: Why GetOk and not Get if you don't use the ok?

File: aci/resource_aci_vnsldevvip.go [hunk 16]
```diff
@@ -0,0 +1,572 @@
...
+	d.SetId(vnsLDevVip.DistinguishedName)
+	vnsLDevVipMap, err := vnsLDevVip.ToMap()
+	if err != nil {
+		return d, err
+	}
+	dn := d.Id()
+	if dn != vnsLDevVip.DistinguishedName {
+		d.Set("tenant_dn", "")
+	} else {
+		d.Set("tenant_dn", GetParentDn(vnsLDevVip.DistinguishedName, fmt.Sprintf("/lDevVip-%s", vnsLDevVipMap["name"])))
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:204]: "/"+models.RnvnsLDevVip

File: aci/resource_aci_vnsldevvip.go [hunk 17]
```diff
@@ -0,0 +1,572 @@
...
+		for _, relationParam := range relationParamMap {
+			inner_map := relationParam.(map[string]interface{})
+			checkDns = append(checkDns, fmt.Sprintf("%v", inner_map["domain_dn"]))
+		}
+	}
+
+	if relationTovnsRsALDevToPhysDomP, ok := d.GetOk("relation_vns_rs_al_dev_to_phys_dom_p"); ok {
+		relationParam := relationTovnsRsALDevToPhysDomP.(string)
+		checkDns = append(checkDns, relationParam)
+
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:342]: Remove empty line?

File: aci/resource_aci_vnsldevvip.go [hunk 18]
```diff
@@ -0,0 +1,572 @@
...
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	d.Partial(false)
+
+	if relationTovnsRsALDevToDomP, ok := d.GetOk("relation_vns_rs_al_dev_to_dom_p"); ok {
+		relationParamMap := relationTovnsRsALDevToDomP.(*schema.Set).List()
+		var switchingMode, domainDN string
+		for _, relationParam := range relationParamMap {
+			inner_map := relationParam.(map[string]interface{})
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:356]: inner_map -> innerMap

File: aci/resource_aci_vnsldevvip.go [hunk 19]
```diff
@@ -0,0 +1,572 @@
...
+	err := aciClient.Save(vnsLDevVip)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	checkDns := make([]string, 0, 1)
+
+	if relationTovnsRsALDevToDomP, ok := d.GetOk("relation_vns_rs_al_dev_to_dom_p"); ok {
+		relationParamMap := relationTovnsRsALDevToDomP.(*schema.Set).List()
+		for _, relationParam := range relationParamMap {
+			inner_map := relationParam.(map[string]interface{})
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:334]: inner_map -> innerMap

File: aci/resource_aci_vnsldevvip.go [hunk 20]
```diff
@@ -0,0 +1,572 @@
...
+		return diag.FromErr(err)
+	}
+	d.Partial(false)
+
+	if relationTovnsRsALDevToDomP, ok := d.GetOk("relation_vns_rs_al_dev_to_dom_p"); ok {
+		relationParamMap := relationTovnsRsALDevToDomP.(*schema.Set).List()
+		var switchingMode, domainDN string
+		for _, relationParam := range relationParamMap {
+			inner_map := relationParam.(map[string]interface{})
+			switchingMode = fmt.Sprintf("%v", inner_map["switching_mode"])
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:357]: Why using Sprintf and %v and not just casting to string with .(string)?

File: aci/resource_aci_vnsldevvip.go [hunk 21]
```diff
@@ -0,0 +1,572 @@
...
+			domainDN = fmt.Sprintf("%v", inner_map["domain_dn"])
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:358]: Why using Sprintf and %v and not just casting to string with .(string)?

File: aci/resource_aci_vnsldevvip.go [hunk 22]
```diff
@@ -0,0 +1,572 @@
...
+
+		if err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	if relationTovnsRsALDevToPhysDomP, ok := d.GetOk("relation_vns_rs_al_dev_to_phys_dom_p"); ok {
+		relationParam := relationTovnsRsALDevToPhysDomP.(string)
+		err = aciClient.CreateRelationvnsRsALDevToPhysDomP(vnsLDevVip.DistinguishedName, vnsLDevVipAttr.Annotation, relationParam)
+
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:370]: Remove empty line

File: aci/resource_aci_vnsldevvip.go [hunk 23]
```diff
@@ -0,0 +1,572 @@
...
+	if relationTovnsRsALDevToDomP, ok := d.GetOk("relation_vns_rs_al_dev_to_dom_p"); ok {
+		relationParamMap := relationTovnsRsALDevToDomP.(*schema.Set).List()
+		var switchingMode, domainDN string
+		for _, relationParam := range relationParamMap {
+			inner_map := relationParam.(map[string]interface{})
+			switchingMode = fmt.Sprintf("%v", inner_map["switching_mode"])
+			domainDN = fmt.Sprintf("%v", inner_map["domain_dn"])
+		}
+		err = aciClient.CreateRelationvnsRsALDevToDomP(vnsLDevVip.DistinguishedName, vnsLDevVipAttr.Annotation, switchingMode, domainDN)
+
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:361]: Remove empty line

File: aci/resource_aci_vnsldevvip.go [hunk 24]
```diff
@@ -0,0 +1,572 @@
...
+	}
+
+	checkDns := make([]string, 0, 1)
+
+	if d.HasChange("relation_vns_rs_al_dev_to_dom_p") || d.HasChange("annotation") {
+		_, newRel := d.GetChange("relation_vns_rs_al_dev_to_dom_p")
+		newRelSet := newRel.(*schema.Set).List()
+		for _, relationParam := range newRelSet {
+			inner_map := relationParam.(map[string]interface{})
+			checkDns = append(checkDns, fmt.Sprintf("%v", inner_map["domain_dn"]))
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:459]: Why using Sprintf and %v and not just casting to string with .(string)?

File: aci/resource_aci_vnsldevvip.go [hunk 25]
```diff
@@ -0,0 +1,572 @@
...
+		return diag.FromErr(err)
+	}
+
+	checkDns := make([]string, 0, 1)
+
+	if d.HasChange("relation_vns_rs_al_dev_to_dom_p") || d.HasChange("annotation") {
+		_, newRel := d.GetChange("relation_vns_rs_al_dev_to_dom_p")
+		newRelSet := newRel.(*schema.Set).List()
+		for _, relationParam := range newRelSet {
+			inner_map := relationParam.(map[string]interface{})
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:458]: inner_map -> innerMap

File: aci/resource_aci_vnsldevvip.go [hunk 26]
```diff
@@ -0,0 +1,572 @@
...
+		return diag.FromErr(err)
+	}
+	d.Partial(false)
+
+	if d.HasChange("relation_vns_rs_al_dev_to_dom_p") || d.HasChange("annotation") {
+		_, newRelParam := d.GetChange("relation_vns_rs_al_dev_to_dom_p")
+		newRelParamMap := newRelParam.(*schema.Set).List()
+		var switchingMode, domainDN string
+		for _, newRelationParam := range newRelParamMap {
+			inner_map := newRelationParam.(map[string]interface{})
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:481]: inner_map -> innerMap

File: aci/resource_aci_vnsldevvip.go [hunk 27]
```diff
@@ -0,0 +1,572 @@
...
+			switchingMode = fmt.Sprintf("%v", inner_map["switching_mode"])
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:482]: Why using Sprintf and %v and not just casting to string with .(string)?

File: aci/resource_aci_vnsldevvip.go [hunk 28]
```diff
@@ -0,0 +1,572 @@
...
+			domainDN = fmt.Sprintf("%v", inner_map["domain_dn"])
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:483]: Why using Sprintf and %v and not just casting to string with .(string)?

File: aci/resource_aci_vnsldevvip.go [hunk 29]
```diff
@@ -0,0 +1,572 @@
...
+		}
+		err = aciClient.DeleteRelationvnsRsALDevToDomP(vnsLDevVip.DistinguishedName)
+
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:486]: Remove empty line

File: aci/resource_aci_vnsldevvip.go [hunk 30]
```diff
@@ -0,0 +1,572 @@
...
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		err = aciClient.CreateRelationvnsRsALDevToDomP(vnsLDevVip.DistinguishedName, vnsLDevVipAttr.Annotation, switchingMode, domainDN)
+
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:491]: Remove empty line

File: aci/resource_aci_vnsldevvip.go [hunk 31]
```diff
@@ -0,0 +1,572 @@
...
+		}
+	}
+	if d.HasChange("relation_vns_rs_al_dev_to_phys_dom_p") || d.HasChange("annotation") {
+		_, newRelParam := d.GetChange("relation_vns_rs_al_dev_to_phys_dom_p")
+		err = aciClient.DeleteRelationvnsRsALDevToPhysDomP(vnsLDevVip.DistinguishedName)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		err = aciClient.CreateRelationvnsRsALDevToPhysDomP(vnsLDevVip.DistinguishedName, vnsLDevVipAttr.Annotation, newRelParam.(string))
+
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:503]: Remove empty line

File: website/docs/d/l4_l7_devices.html.markdown [hunk 32]
```diff
@@ -0,0 +1,55 @@
...
+```hcl
+data "aci_l4_l7_devices" "example" {
+  tenant_dn  = aci_tenant.example.id
+  name       = "example"
+}
+```
+
+## Argument Reference ##
+
+* `tenant_dn` - (Required) Distinguished name of parent Tenant object.
```
Comment (lhercot) [inline website/docs/d/l4_l7_devices.html.markdown:36]: of the parent Tenant object

File: website/docs/d/l4_l7_devices.html.markdown [hunk 33]
```diff
@@ -0,0 +1,55 @@
...
+* `name` - (Required) Name of object L4-L7 Devices.
```
Comment (lhercot) [inline website/docs/d/l4_l7_devices.html.markdown:37]: Name of the L4-L7 Device.

File: website/docs/d/l4_l7_devices.html.markdown [hunk 34]
```diff
@@ -0,0 +1,55 @@
...
+
+## Attribute Reference ##
+* `id` - Attribute id set to the Dn of the L4-L7 Devices.
```
Comment (lhercot) [inline website/docs/d/l4_l7_devices.html.markdown:40]: Devices -> Device

File: website/docs/d/l4_l7_devices.html.markdown [hunk 35]
```diff
@@ -0,0 +1,55 @@
...
+* `annotation` - (Optional) Annotation of object L4-L7 Devices.
```
Comment (lhercot) [inline website/docs/d/l4_l7_devices.html.markdown:41]: of the L4-L7 Device object.

File: website/docs/d/l4_l7_devices.html.markdown [hunk 36]
```diff
@@ -0,0 +1,55 @@
...
+* `name_alias` - (Optional) Name Alias of object L4-L7 Devices.
```
Comment (lhercot) [inline website/docs/d/l4_l7_devices.html.markdown:42]: of the L4-L7 Device object.

File: website/docs/d/l4_l7_devices.html.markdown [hunk 37]
```diff
@@ -0,0 +1,55 @@
...
+* `active` - (Optional) Active mode. Allowed values are "no", "yes", and default value is "no". Type: String.
```
Comment (lhercot) [inline website/docs/d/l4_l7_devices.html.markdown:43]: What is the active mode used for? Can we explain that a bit better?

File: website/docs/d/l4_l7_devices.html.markdown [hunk 38]
```diff
@@ -0,0 +1,55 @@
...
+* `context_aware` - (Optional) Determines if the L4-L7 device cluster supports multiple contexts (VRFs). Allowed values are "multi-Context", "single-Context", and default value is "single-Context". Type: String.
+* `device_type` - (Optional) Device Type. Allowed values are "CLOUD", "PHYSICAL", "VIRTUAL", and default value is "PHYSICAL". Type: String.
+* `function_type` - (Optional) Function Type of the L4-L7 device cluster. Allowed values are "GoThrough", "GoTo", "L1", "L2", "None", and default value is "GoTo". Type: String.
+* `is_copy` - (Optional) Sets device as a copy. Allowed values are "no", "yes", and default value is "no". Type: String.
+* `mode` - (Optional) Specifies if the device is legacy (classical VLAN/VXLAN) or supports service tag switching (STS). Allowed value is "legacy-Mode", and default value is "legacy-Mode". Type: String.
```
Comment (lhercot) [inline website/docs/d/l4_l7_devices.html.markdown:48]: Should we expose this if it is the default and the only option?

File: website/docs/d/l4_l7_devices.html.markdown [hunk 39]
```diff
@@ -0,0 +1,55 @@
...
+* `promiscuous_mode` - (Optional) Promiscuous Mode supports port groups in an external VMM controller, such as a Vcenter. This needs to be turned on only for service devices in the cloud, not for Enterprise AVE service deployments. Allowed values are "no", "yes", and default value is "no". Type: String.
```
Comment (lhercot) [inline website/docs/d/l4_l7_devices.html.markdown:49]: This explanation is not clear. Also use vCenter not Vcenter.

File: website/docs/d/l4_l7_devices.html.markdown [hunk 40]
```diff
@@ -0,0 +1,55 @@
...
+* `service_type` - (Optional) Service Type. Allowed values are "ADC", "COPY", "FW", "NATIVELB", "OTHERS", and default value is "OTHERS". Type: String.
```
Comment (lhercot) [inline website/docs/d/l4_l7_devices.html.markdown:50]: (Optional) The type of service the L4-L7 device performs. Allowed...

File: website/docs/d/l4_l7_devices.html.markdown [hunk 41]
```diff
@@ -0,0 +1,55 @@
...
+* `trunking` - (Optional) Sets trunking port group for virtual devices. Allowed values are "no", "yes", and default value is "no". Type: String.
```
Comment (lhercot) [inline website/docs/d/l4_l7_devices.html.markdown:51]: Configures the device port group for trunking for virtual devices.

File: website/docs/d/l4_l7_devices.html.markdown [hunk 42]
```diff
@@ -0,0 +1,55 @@
...
+* `relation_vns_rs_al_dev_to_dom_p` - (Optional) Represents a relation from L4-L7 Device to a Vmm Domain Profile (class vmmDomP). Type: Block.
```
Comment (lhercot) [inline website/docs/d/l4_l7_devices.html.markdown:52]: VMM

File: website/docs/d/l4_l7_devices.html.markdown [hunk 43]
```diff
@@ -0,0 +1,55 @@
...
+  * `domain_dn` - (Required) Distinguished name of the target. Type: String.
```
Comment (lhercot) [inline website/docs/d/l4_l7_devices.html.markdown:53]: Distinguished name of the VMM Domain in which the VM is deployed.

File: website/docs/d/l4_l7_devices.html.markdown [hunk 44]
```diff
@@ -0,0 +1,55 @@
...
+  * `switching_mode` - (Optional) Switching mode. Allowed values are "native", "AVE", and default value is "native". The value "AVE" is not supported with non-AVE VMM Domain. Type: String.
```
Comment (lhercot) [inline website/docs/d/l4_l7_devices.html.markdown:54]: Port group switching mode.

File: website/docs/d/l4_l7_devices.html.markdown [hunk 45]
```diff
@@ -0,0 +1,55 @@
+---
+layout: "aci"
```
Comment (lhercot) [inline website/docs/d/l4_l7_devices.html.markdown:2]: missing subcategory: "L4-L7 Services"

File: website/docs/d/l4_l7_devices.html.markdown [hunk 46]
```diff
@@ -0,0 +1,55 @@
...
+page_title: "ACI: aci_l4_l7_devices"
+sidebar_current: "docs-aci-data-source-l4_l7_devices"
+description: |-
+  Data source for ACI L4-L7 Devices
+---
+
+# aci_l4_l7_devices #
+
+Data source for ACI L4-L7 Devices
```
Comment (lhercot) [inline website/docs/d/l4_l7_devices.html.markdown:11]: Device

File: website/docs/r/l4_l7_devices.html.markdown [hunk 47]
```diff
@@ -0,0 +1,92 @@
+---
+layout: "aci"
```
Comment (lhercot) [inline website/docs/r/l4_l7_devices.html.markdown:2]: Missing subcategory: "L4-L7 Services"

File: website/docs/r/l4_l7_devices.html.markdown [hunk 48]
```diff
@@ -0,0 +1,92 @@
...
+page_title: "ACI: aci_l4_l7_devices"
+sidebar_current: "docs-aci-resource-l4_l7_devices"
+description: |-
+  Manages ACI L4-L7 Devices
+---
+
+# aci_l4_l7_devices #
+
+Manages ACI L4-L7 Devices
```
Comment (lhercot) [inline website/docs/r/l4_l7_devices.html.markdown:11]: Device

File: website/docs/r/l4_l7_devices.html.markdown [hunk 49]
```diff
@@ -0,0 +1,92 @@
...
+
+
+## Example Usage ##
+
+### Creating a PHYSICAL Device ###
+
+```hcl
+resource "aci_l4_l7_devices" "physical_device" {
+  tenant_dn                            = aci_tenant.terraform_tenant.id
+  name                                 = "example2"
```
Comment (lhercot) [inline website/docs/r/l4_l7_devices.html.markdown:30]: example1

File: website/docs/r/l4_l7_devices.html.markdown [hunk 50]
```diff
@@ -0,0 +1,92 @@
...
+  relation_vns_rs_al_dev_to_phys_dom_p = "uni/phys-test_dom"
+}
+```
+
+### Creating a VIRTUAL Device ###
+
+```hcl
+resource "aci_l4_l7_devices" "virtual_device" {
+  tenant_dn        = aci_tenant.terraform_tenant.id
+  name             = "example1"
```
Comment (lhercot) [inline website/docs/r/l4_l7_devices.html.markdown:48]: example2

File: website/docs/r/l4_l7_devices.html.markdown [hunk 51]
```diff
@@ -0,0 +1,92 @@
...
+  trunking         = "no"
+  relation_vns_rs_al_dev_to_dom_p {
+    target_dn      = "uni/vmmp-VMware/dom-ESX0-leaf102-vds"
+    switching_mode = "AVE"
+  }
+}
+```
+
+## Argument Reference ##
+
```
Comment (lhercot) [inline website/docs/r/l4_l7_devices.html.markdown:66]: Apply same comments as datasource.

**Review by lhercot — CHANGES_REQUESTED** [review]

File: website/docs/d/l4_l7_devices.html.markdown (hunk 39 above)
Comment (shrsr) [reply website/docs/d/l4_l7_devices.html.markdown:49]: I got the explanation from UI.

File: aci/data_source_aci_vnsldevvip.go [hunk 52]
```diff
@@ -0,0 +1,158 @@
...
+				Type:          schema.TypeSet,
+				Optional:      true,
+				Description:   "Create relation to vmmDomP",
+				MaxItems:      1,
+				ConflictsWith: []string{"relation_vns_rs_al_dev_to_phys_dom_p"},
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"domain_dn": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
```
Comment (lhercot) [inline aci/data_source_aci_vnsldevvip.go:93]: This one shouldn't be required in the datasource right? Should be optional and computed right?

File: aci/data_source_aci_vnsldevvip.go [hunk 53]
```diff
@@ -0,0 +1,158 @@
...
+								"AVE",
+								"native",
+							}, false),
+						},
+					},
+				},
+			},
+			"relation_vns_rs_al_dev_to_phys_dom_p": {
+				Type:        schema.TypeString,
+				Optional:    true,
```
Comment (lhercot) [inline aci/data_source_aci_vnsldevvip.go:108]: computed?

File: aci/data_source_aci_vnsldevvip.go [hunk 54]
```diff
@@ -0,0 +1,158 @@
...
+				ConflictsWith: []string{"relation_vns_rs_al_dev_to_phys_dom_p"},
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"domain_dn": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"switching_mode": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
```
Comment (lhercot) [inline aci/data_source_aci_vnsldevvip.go:97]: computed?

File: aci/data_source_aci_vnsldevvip.go [hunk 55]
```diff
@@ -0,0 +1,158 @@
...
+				Computed: true,
+			},
+			"trunking": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"relation_vns_rs_al_dev_to_dom_p": {
+				Type:          schema.TypeSet,
+				Optional:      true,
```
Comment (lhercot) [inline aci/data_source_aci_vnsldevvip.go:85]: computed?

File: aci/resource_aci_vnsldevvip.go [hunk 56]
```diff
@@ -0,0 +1,566 @@
...
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"domain_dn": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"switching_mode": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							Default:  "native",
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:151]: Do you have to force a default here?

File: aci/resource_aci_vnsldevvip.go [hunk 57]
```diff
@@ -0,0 +1,566 @@
...
+								"AVE",
+								"native",
+							}, false),
+						},
+					},
+				},
+			},
+			"relation_vns_rs_al_dev_to_phys_dom_p": {
+				Type:        schema.TypeString,
+				Optional:    true,
```
Comment (lhercot) [inline aci/resource_aci_vnsldevvip.go:162]: Computed?

File: website/docs/d/l4_l7_devices.html.markdown [hunk 58]
```diff
@@ -0,0 +1,55 @@
...
+
+## Attribute Reference ##
+* `id` - Attribute id set to the Dn of the L4-L7 Device.
+* `annotation` - (Optional) Annotation of the L4-L7 Device object.
+* `name_alias` - (Optional) Name Alias of the L4-L7 Device object.
+* `active` - (Optional) Enables L4-L7 device cluster to operate in active/active mode. Allowed values are "no", "yes", and default value is "no". Type: String.
+* `context_aware` - (Optional) Determines if the L4-L7 device cluster supports multiple contexts (VRFs). Allowed values are "multi-Context", "single-Context", and default value is "single-Context". Type: String.
+* `device_type` - (Optional) Device Type. Allowed values are "CLOUD", "PHYSICAL", "VIRTUAL", and default value is "PHYSICAL". Type: String.
+* `function_type` - (Optional) Function Type of the L4-L7 device cluster. Allowed values are "GoThrough", "GoTo", "L1", "L2", "None", and default value is "GoTo". Type: String.
+* `is_copy` - (Optional) Sets device as a copy. Allowed values are "no", "yes", and default value is "no". Type: String.
```
Comment (lhercot) [inline website/docs/d/l4_l7_devices.html.markdown:48]: Sets device as a copy device.

File: website/docs/d/l4_l7_devices.html.markdown [hunk 59]
```diff
@@ -0,0 +1,55 @@
...
+* `promiscuous_mode` - (Optional) Enabling Promiscuous Mode supports all the traffic in a port group to reach a VM attached to a promiscuous port. Allowed values are "no", "yes", and default value is "no". Type: String.
```
Comment (lhercot) [inline website/docs/d/l4_l7_devices.html.markdown:49]: Enabling Promiscuous Mode to allow all the traffic...

File: website/docs/d/l4_l7_devices.html.markdown [hunk 60]
```diff
@@ -0,0 +1,55 @@
...
+* `service_type` - (Optional) The type of service the L4-L7 device performs. Allowed values are "ADC", "COPY", "FW", "NATIVELB", "OTHERS", and default value is "OTHERS". Type: String.
+* `trunking` - (Optional) Configures the device port group for trunking of virtual devices. Allowed values are "no", "yes", and default value is "no". Type: String.
+* `relation_vns_rs_al_dev_to_dom_p` - (Optional) Represents a relation from L4-L7 Device to a VMM Domain Profile (class vmmDomP). Type: Block.
+  * `domain_dn` - (Required) Distinguished name of the VMM Domain in which the VM is deployed. Type: String.
```
Comment (lhercot) [inline website/docs/d/l4_l7_devices.html.markdown:53]: This shouldn't be required.

File: website/docs/r/l4_l7_devices.html.markdown [hunk 61]
```diff
@@ -0,0 +1,92 @@
...
+* `tenant_dn` - (Required) Distinguished name of the parent Tenant object.
+* `name` - (Required) Name of the L4-L7 Device.
+* `annotation` - (Optional) Annotation of the L4-L7 Device object.
+* `active` - (Optional) Enables L4-L7 device cluster to operate in active/active mode. Allowed values are "no", "yes", and default value is "no". Type: String.
+* `context_aware` - (Optional) Determines if the L4-L7 device cluster supports multiple contexts (VRFs). Allowed values are "multi-Context", "single-Context", and default value is "single-Context". Type: String.
+* `device_type` - (Optional) Device Type. Allowed values are "CLOUD", "PHYSICAL", "VIRTUAL", and default value is "PHYSICAL". Type: String.
+* `function_type` - (Optional) Function Type of the L4-L7 device cluster. Allowed values are "GoThrough", "GoTo", "L1", "L2", "None", and default value is "GoTo". Type: String.
+* `is_copy` - (Optional) Sets device as a copy. Allowed values are "no", "yes", and default value is "no". Type: String.
```
Comment (lhercot) [inline website/docs/r/l4_l7_devices.html.markdown:75]: as a copy device

File: website/docs/r/l4_l7_devices.html.markdown [hunk 62]
```diff
@@ -0,0 +1,92 @@
...
+* `promiscuous_mode` - (Optional) Enabling Promiscuous Mode supports all the traffic in a port group to reach a VM attached to a promiscuous port. Allowed values are "no", "yes", and default value is "no". Type: String.
```
Comment (lhercot) [inline website/docs/r/l4_l7_devices.html.markdown:76]: Enabling Promiscuous Mode to allow all the traffic...

**Review by lhercot — CHANGES_REQUESTED** [review]

File: aci/data_source_aci_vnsldevvip.go (hunk 52 above)
Comment (shrsr) [reply aci/data_source_aci_vnsldevvip.go:93]: Changed it to optional

**Review by lhercot — APPROVED** [review]
LGTM

**Pull request merged**